Clarify semantics of the various pointer to pointer casts (#1451)
rust-lang/reference: src/expressions/operator-expr.md

assert_eq!(values[1], 3);
```

#### Pointer-to-pointer cast

`*const T` / `*mut T` can be cast to `*const U` / `*mut U` with the following behavior:

> - If `T` is unsized and `U` is sized, the cast discards all metadata that

Reviewer: ```suggestion
- If `T` and `U` are both sized, the pointer is returned unchanged.
- If `T` and `U` are both unsized, the pointer is also returned unchanged. In particular,
  the metadata is preserved exactly.

    For instance, a cast from `*const [T]` to `*const [U]` preserves the number of elements.
  Note that, as a consequence, such casts do not necessarily preserve the size of the
  pointer's referent (e.g., casting `*const [u16]` to `*const [u8]` will result in a raw
  pointer which refers to an object of half the size of the original). The same
  holds for `str` and any compound type whose unsized tail is a slice type, such
  as `struct Foo(i32, [u8])` or `(u64, Foo)`.
- If `T` is unsized and `U` is sized, the cast discards all metadata that
```

> - If `T` is unsized and `U` is sized, the cast discards all metadata that
>   completes the wide pointer `T` and produces a thin pointer `U` to the
>   value of the unsized pointer.

Reviewer: ```suggestion
- If `T` is unsized and `U` is sized, the cast discards all metadata that
  completes the wide pointer `T` and produces a thin pointer `U` consisting
  of the data part of the unsized pointer.
```

assert_eq!(values[1], 3);
```

#### Pointer-to-pointer cast

`*const T` / `*mut T` can be cast to `*const U` / `*mut U` with the following behavior:

- If `T` and `U` are both sized, the pointer is returned unchanged.
- If `T` and `U` are both unsized, the pointer is also returned unchanged. In particular,
  the metadata is preserved exactly.

    For instance, a cast from `*const [T]` to `*const [U]` preserves the number of elements.
  Note that, as a consequence, such casts do not necessarily preserve the size of the
  pointer's referent (e.g., casting `*const [u16]` to `*const [u8]` will result in a raw
  pointer which refers to an object of half the size of the original). The same
  holds for `str` and any compound type whose unsized tail is a slice type, such
  as `struct Foo(i32, [u8])` or `(u64, Foo)`.
- If `T` is unsized and `U` is sized, the cast discards all metadata that
  completes the wide pointer `T` and produces a thin pointer `U` consisting
  of the data part of the unsized pointer.

> - Otherwise a cast from `T` to `U` produces a value identical to the original
>   pointer.

Reviewer: ```suggestion
```
This is now covered by the first two cases.

assert_eq!(values[1], 3);
```

#### Sized to sized pointer casts

A sized pointer to pointer cast `*const T as *const U` or `*mut T as *mut U`
where `T: Sized` and `U: Sized` produces a value identical to the original

> pointer.

Reviewer: We also allow const-to-mut and mut-to-const

Author: What's a clear way to enumerate all permutations concisely?

Author: Saw the last comment.

#### Unsized to unsized pointer casts

Any unsized to unsized pointer cast `*const T as *const U` or `*mut T as *mut U`
produces a value identical to the original pointer if `T` and `U` are compatible

> unsized types.

Reviewer: This then raises the question, what does it do if they are not compatible?

Author: Do we have to answer that now? I'd like to avoid pulling as many threads as possible. 😅

Reviewer: The PR in its current form introduces a confusing new sentence. So yes this needs to be fixed.

But I think the structure I propose will also handle this pretty naturally.

Author: Is the sentence "compatible unsized types"? If so, that's not my phrasing. It's borrowed from the table just above.

Reviewer: Ah, I see. But in the table it is clear that if the types are not compatible, then the cast is simply not permitted.

Author: Ok, so I've removed the double reference to validity since it's already covered by the table and rewrote it to only focus on behavior. Please check it out!

For slice types like `[T]` and `[U]`, the raw pointer types `*const [T]`, `*mut
[T]`, `*const [U]`, and `*mut [U]` encode the number of elements in this slice.
Casts between these raw pointer types preserve the number of elements. Note
that, as a consequence, such casts do *not* necessarily preserve the size of the
pointer's referent (e.g., casting `*const [u16]` to `*const [u8]` will result in
a raw pointer which refers to an object of half the size of the original). The
same holds for `str` and any compound type whose unsized tail is a slice type,
such as struct `Foo(i32, [u8])` or `(u64, Foo)`.

#### Unsized to sized pointer casts

> Any wide to narrow pointer cast discards all metadata that completes the wide

Reviewer 2: The word "metadata" isn't defined anywhere. Can you add a description here that defines what it means?

Author: I don't know if I'm just promoting kicking the can down the road, but I opted not to define metadata since:
* It's not necessary to define the semantics of this cast, the only distinction is that unsized pointers have metadata and this cast removes it.
* Another part of the reference "behaviors considered undefined" uses metadata without defining it.

That being said, if I were to define it, where do you think I should do it and to what extent?

> The representation of pointers and references to DSTs is made up of two components; A base pointer and metadata. The metadata is the necessary extra data that has to be attached to pointers and references of DSTs which completes their representation. This is stored adjacent to the base pointer causing them to be wider than sized pointers earning them the name "wide" or "fat" pointers. The base pointer can be extracted through an unsized to sized pointer cast.

I'm a bit concerned that I'll end up having to define everything metadata adjacent 😓, so at least I'd want to avoid that as much as possible.

pointer and produces a thin base pointer.

The following casts are well-defined:

* `*const [T] to *const U` or `*mut [T] to *mut U` discards the length of the
  slice and produces a pointer to the first element. The produced pointer is
  casted to `U` in the same manner as a sized pointer to pointer cast.
* `*const dyn T to *const U` or `*mut dyn T to *mut U` discards all metadata
  associated with the [trait object] and produces a pointer to the instance from
  which the trait object was constructed. The produced pointer is casted to `U`

>   in the same manner as a sized pointer to pointer cast.

Reviewer: This list is not exhaustive, e.g. it misses this case:
```rust
fn test(x: *const (u8, str)) -> *const u16 { x as _ }
```
Also, it seems strange to list both of these only to then say they do basically the same. The enumeration just kind of repeats what the text above says?

Author: So you either want to also enumerate unsized tuples and presumably unsized custom types or collapse all of them into one?

It would be nice to list the common instances (`*const [T]` / `*const dyn T`) somehow, but maybe not necessary?

I would like to defer to your preference on this matter, anything works for me as long as the unsized pointer cast semantics are documented.

Author: Ah, saw the last comment now.

Reviewer: > It would be nice to list the common instances (*const [T] / *const dyn T) somehow, but maybe not necessary?

This is a reference, not tutorial material. So no, common cases should not be listed. Every special case listed should only be listed because it actually is a special case that behaves different from the general case.

Reviewer: Fair. If you feel it's useful, add something like that -- but the "for example" wording is crucial here.

However, I'm not convinced that an example is actually helpful here. This will be easier to discuss once there is concrete text for the revised version. :)

`*const T` / `*mut T` can be cast to `*const U` / `*mut U` with the following behavior:

- If `T` is unsized and `U` is sized, the cast discards all metadata that
  completes the wide pointer `T` and produces a thin pointer `U` to the

>   value of the unsized pointer.

Author: I'm struggling to find a phrasing here for to describe what "value" is which applies to all unsized pointers and also conveys *what exactly it points to*. In the reference for trait objects "instance" is used, but that would be a bit unorthodox for slices.
